ehrapy/api/preprocessing/_data_imputation.py

```diff
@@ -3,51 +3,300 @@
+        # impute all columns if None passed
+        else:
+            adata.X = imputer.fit_transform(adata.X)
+    # Imputation using median strategy only works with numerical data
+    except ValueError:
+        raise ImputeStrategyNotAvailableError(
+            "Can only impute numerical data using median strategy. Try to restrict imputation"
+            "to certain columns using replacement parameter or use a different mode."
+        )
+
+    return adata
+
+
+def _most_frequent(adata: AnnData, **kwargs) -> AnnData:
+    """
+    Impute AnnData object using most_frequent imputation. This works for both, numerical and non-numerical data.
+
+    Args:
+        adata: The AnnData object to use most_frequent Imputation on
+        **kwargs: Keyword args
+                  replacement: A list of strings indicating which columns to use most_frequent imputation on (if None -> all columns)
+
+    Returns:
+            The imputed AnnData object
+
+    """
+    replacement = kwargs.get("replacement")
+    imputer = SimpleImputer(strategy="most_frequent")
+    # impute a subset of columns
+    if isinstance(replacement, list):
+        column_indices = get_column_indices(adata, replacement)
+        adata.X[::, column_indices] = imputer.fit_transform(adata.X[::, column_indices])
+    # impute all columns if None passed
+    else:
+        adata.X = imputer.fit_transform(adata.X)
+
+    return adata
+
+
+# ===================== KNN Imputation =========================
+
+
+def _knn(adata: AnnData, **kwargs) -> AnnData:
+    """Impute data using the KNN-Imputer.
+    When using KNN Imputation with mixed data (non-numerical and numerical), encoding using ordinal encoding is required
+    since KNN Imputation can only work on numerical data. The encoding itself is just a utility and will be redone once
+    imputation ran successfully.
+
+    Args:
+        adata: The AnnData object to use KNN Imputation on
+        **kwargs: Keyword args
+                  replacement: A list of strings indicating which columns to use KNN imputation on (if None -> all columns)
+
+    Returns:
+            The imputed (but unencoded) AnnData object
+    """
+    # numerical only data needs no encoding since KNN Imputation can be applied directly
+    if np.issubdtype(adata.X.dtype, np.number):
+        _knn_impute(adata, **kwargs)
+    else:
+        # ordinal encoding is used since non-numerical data can not be imputed using KNN Imputation
+        enc = OrdinalEncoder()
+        adata.X = enc.fit_transform(adata.X)
+        # impute the data using KNN imputation
+        _knn_impute(adata, **kwargs)
+        # decode ordinal encoding to obtain imputed original data
+        adata.X = enc.inverse_transform(adata.X)
+
+    return adata
+
+
+def _knn_impute(adata: AnnData, **kwargs) -> None:
+    """Utility function to impute data using KNN-Imputer"""
+    n_neighbors = kwargs.get("n_neighbors")
```

Comment:
@Zethson ; might be better to remove this one. Especially for imputing non-numerical data using KNN its important to only use the next nearest neighbor, instead of rendering a meaningless mean of some values, that could not be decoded afterwards. Thoughts?

Reply:
Let's discuss today.